src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs

```diff
@@ -298,154 +161,156 @@ private void OnRunCodeAnalysisForSelectedProjectStatus(object sender, EventArgs
-            var statusBar = await _statusbar.GetValueOrNullAsync().ConfigureAwait(true);
-            var totalProjectCount = project != null ? (1 + otherProjectsForMultiTfmProject.Length) : solution.ProjectIds.Count;
-            using var statusBarUpdater = statusBar != null
-                ? new StatusBarUpdater(statusBar, _threadingContext, projectOrSolutionName, (uint)totalProjectCount)
-                : null;
-
-            await TaskScheduler.Default;
+            try
+            {
+                using var asyncToken = _listener.BeginAsyncOperation($"{nameof(VisualStudioDiagnosticAnalyzerService)}_{nameof(RunAnalyzers)}");
 
-            var onAfterProjectAnalyzed = statusBarUpdater != null ? statusBarUpdater.OnAfterProjectAnalyzed : (Action<Project>)((Project _) => { });
-            await _codeAnalysisService.RunAnalysisAsync(solution, project?.Id, onAfterProjectAnalyzed, CancellationToken.None).ConfigureAwait(false);
+                // Add a message to VS status bar that we are running code analysis.
+                using var statusBarUpdater = new StatusBarUpdater(
+                    this, await _statusbar.GetValueOrNullAsync(cancellationToken).ConfigureAwait(true),
+                    progressName, totalProjectCount: projectsToAnalyze.Length, cancellationToken);
 
-            foreach (var otherProject in otherProjectsForMultiTfmProject)
-                await _codeAnalysisService.RunAnalysisAsync(solution, otherProject.Id, onAfterProjectAnalyzed, CancellationToken.None).ConfigureAwait(false);
+                await RoslynParallel.ForEachAsync(
+                    projectsToAnalyze,
+                    cancellationToken,
+                    async (project, cancellationToken) =>
+                    {
+                        try
+                        {
+                            await _codeAnalysisService.RunAnalysisAsync(project, cancellationToken).ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            statusBarUpdater.OnAfterProjectAnalyzed();
+                        }
+                    }).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+            catch (Exception ex) when (FatalError.ReportAndCatch(ex))
+            {
+            }
         });
     }
 
-    private Project? GetProject(IVsHierarchy? hierarchy)
+    private Project? GetProject(Solution solution, IVsHierarchy? hierarchy)
     {
         if (hierarchy != null)
         {
             var projectMap = _workspace.Services.GetRequiredService<IHierarchyItemToProjectIdMap>();
             var projectHierarchyItem = _vsHierarchyItemManager.GetHierarchyItem(hierarchy, VSConstants.VSITEMID_ROOT);
             if (projectMap.TryGetProjectId(projectHierarchyItem, targetFrameworkMoniker: null, out var projectId))
-            {
-                return _workspace.CurrentSolution.GetProject(projectId);
```

Comment:
we had multiple piecews of code accessing _workspace.CurrentSoltion.  This is racey.  Changed to pass the solution snapshot in so that we do things consistently.